README.md
Cognite GraphQL Python Generator
==========================

**THIS PACKAGE HAS BEEN RENAMED** It is now called `cognite-pygen`

* New pypy.org: [https://pypi.org/project/cognite-pygen/](https://pypi.org/project/cognite-pygen/).
* New GitHub: [https://github.com/cognitedata/pygen](https://github.com/cognitedata/pygen)

-------------------------------------------------------------------------------


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Changes are grouped as follows
- `Added` for new features.
- `Changed` for changes in existing functionality.
- `Deprecated` for soon-to-be removed features.
- `Improved` for transparent changes, e.g. better performance.
- `Removed` for now removed features.
- `Fixed` for any bug fixes.
- `Security` in case of vulnerabilities.

## [0.8.1] - 22-05-23

### Changed

* The package name `cognite-gql-pygen` points to `cognite-pygen`.


## [0.8.0] - 14-05-23

### Improved

* Split dependencies into required and optional.
* Removed all unused dependencies.
* Updated documentation to reflect non-CLI usage.


## [0.7.0] - 13-05-23

### Fix

* Less strict versioning of other packages to support usage in streamlit pyodide runtime.

## [0.6.0] - 13-05-23

### Fix

* Less strict `packaging` versioning (>=21) to support usage in streamlit pyodide runtime.

## [0.5.0] - 10-05-23

### Added

* Testing Client

### Fix

* Issue when the output folder is not relative to the current working directory.


## [0.4.5] - 10-05-23

### Fixed

* Support for nested types
* Typo for `Int` built-in type, was incorrecly `Integer`

### Improved

* Items that are references are now showing this in __repr__ and __str__.


## [0.4.4] - 09-05-23

### Added

* Added compatibility with cognite-sdk 6.x.x


## [0.4.3] - 05-05-23

### Fixed

* Set the license of the package in poetry build.


## [0.4.2] - 05-05-23

### Added

* Support for basic built-in types for the parser (`bool`, `int`, `float`, with `str` from before)


## [0.4.1] - 03-05-23

### Added

* Support for adding relationships via `.connect`

### Changed

* Optimised use of cache internally.


## [0.4.0] - 26-04-23

### Added

* Support for `pip install -e` for local development.

### Changed

* In the CLI, all commands are not under one `dm` tool.

### Removed

* All bash scripts have been removed (under `dm_clinents/bin`).
* CLI tool `dm_clients` was removed (functions moved to `dm`).

### Fixed

* Added a lock around usages of non-thread-safe cache.


## [0.3.0] - 24-04-23

### Fixed

* In the CLI, `dm topython` the argument `name` is now a option instead of a positional argument. This matched the
  documentation in the README.

### Changed

* Renamed `DomainModelAPI.create` to `DomainModelAPI.apply`, to reflect the usage of the underlying endpoint.


## [0.2.1] - 24-04-23

### Fixed

* Fixed a problem with reading annotations caused by reverse ordering of dataclasses (forward refs).


## [0.2.0] - 19-04-23

### Added

* Support for converting from `.graphql` to `pydantic`. This means you can now write your types in a `.graphql` file
  and automatically generate the `pytandic` counterpart.
* CLI for doing the conversion between GraphQL and `pydantic`. The following two commands
  * `dm togql` converts a pydantic schema to `.graphql` and creates the `client.py` which is the SDK to interact
    with the data model through Python.
  * `dm topython` converts a `.graphql` to `pytandic` classes. This command also creates the `client.py`, i.e., the
    SDK to interact with the data model through Pytho
* Documentation of motivation and usage of package.

### Changed

* CLI command going from `gqlpygen` to `pygen`. This is for ease of use.


## [0.1.4]

Changed configuration / settings from a static `config.yaml` to dynaconf and `settings.toml`.

## [0.1.3]

Changed package name from `fdm` to `dm_clients`.

## [0.1.2]

Added support for `Timestamp` and `JSONObject` types.

## [0.1.1]

Added `dm_clients` package.

## [0.1.0]

Initial commit.


cognite/fdm/scalars/timestamp.py
from __future__ import annotations

import logging
import re
from contextlib import suppress
from datetime import datetime
from typing import TYPE_CHECKING, Annotated, TypeVar

logger = logging.getLogger(__name__)


__all__ = [
    "Timestamp",
]


TimestampParsable = TypeVar("TimestampParsable", str, datetime)


class _Timestamp(str):
    """
    String in the format 'YYYY-MM-DDTHH:MM:SS[.millis][Z|time zone]' with optional milliseconds having precision
    of 1-3 decimal digits and optional timezone with format ±HH:MM, ±HHMM, ±HH or Z, where Z represents UTC, Year
    must be between 0001 and 9999.

    This class is a custom Pydantic type:

    >>> from pydantic import BaseModel
    >>> class Foo(BaseModel):
    ...     bar: Timestamp

    # timezones:
    >>> Foo(bar="2023-04-05T06:07:08")
    Foo(bar=Timestamp('2023-04-05T06:07:08'))
    >>> Foo(bar="2023-04-05T06:07:08+09:00")
    Foo(bar=Timestamp('2023-04-05T06:07:08+09:00'))
    >>> Foo(bar="2023-04-05T06:07:08+0900")
    Foo(bar=Timestamp('2023-04-05T06:07:08+09:00'))
    >>> Foo(bar="2023-04-05T06:07:08+09")
    Foo(bar=Timestamp('2023-04-05T06:07:08+09:00'))
    >>> Foo(bar="2023-04-05T06:07:08Z")
    Foo(bar=Timestamp('2023-04-05T06:07:08+00:00'))

    # milliseconds:
    >>> Foo(bar="2023-04-05T06:07:08.123456")
    Foo(bar=Timestamp('2023-04-05T06:07:08.123'))
    >>> Foo(bar="2023-04-05T06:07:08.1256Z")
    Foo(bar=Timestamp('2023-04-05T06:07:08.126+00:00'))
    >>> Foo(bar="2023-04-05T06:07:08.12Z")
    Foo(bar=Timestamp('2023-04-05T06:07:08.120+00:00'))

    # edge years:
    >>> Foo(bar="9999-04-05T06:07:08.12Z")
    Foo(bar=Timestamp('9999-04-05T06:07:08.120+00:00'))
    >>> Foo(bar="0001-04-05T06:07:08.12Z")
    Foo(bar=Timestamp('0001-04-05T06:07:08.120+00:00'))

    # pass in datetime objects:
    >>> from datetime import datetime, timedelta, timezone
    >>> Foo(bar=datetime(2023, 4, 5, 6, 7, 8))
    Foo(bar=Timestamp('2023-04-05T06:07:08.000'))
    >>> Foo(bar=datetime(2023, 4, 5, 6, 7, 8, 123456))
    Foo(bar=Timestamp('2023-04-05T06:07:08.123'))
    >>> Foo(bar=datetime(2023, 4, 5, 6, 7, 8, 345678))
    Foo(bar=Timestamp('2023-04-05T06:07:08.346'))
    >>> Foo(bar=datetime(2023, 4, 5, 6, 7, 8, 345678, timezone(timedelta(hours=-5), 'EST')))
    Foo(bar=Timestamp('2023-04-05T06:07:08.346-05:00'))
    """

    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def __modify_schema__(cls, field_schema):
        field_schema["description"] = (
            "String in the format 'YYYY-MM-DDTHH:MM:SS[.millis][Z|time zone]' with optional milliseconds having"
            " precision of 1-3 decimal digits and optional timezone with format ±HH:MM, ±HHMM, ±HH or Z, where Z"
            " represents UTC, Year must be between 0001 and 9999."
        )

    @classmethod
    def validate(cls, value: TimestampParsable) -> _Timestamp:
        if isinstance(value, str):
            # Add minutes to the timezone offset if missing
            if re.match(r".*[+-]\d\d$", value):
                value = f"{value}00"

            formats = (
                "%Y-%m-%dT%H:%M:%S.%f%z",
                "%Y-%m-%dT%H:%M:%S%z",
                "%Y-%m-%dT%H:%M:%S.%f",
                "%Y-%m-%dT%H:%M:%S",
            )

            def try_parse():
                for fmt in formats:
                    with suppress(ValueError):
                        return datetime.strptime(value, fmt), fmt
                raise ValueError(f"Cannot parse Timestamp: {value}")

            parsed_value, matched_format = try_parse()
            is_subsecond = ".%f" in matched_format

        else:
            parsed_value = value  # assuming datetime
            is_subsecond = True  # cannot distinguish between None and 0 for microseconds

        # Milliseconds need some special treatment:
        microseconds = parsed_value.microsecond
        str_value = parsed_value.replace(microsecond=0).isoformat()
        if is_subsecond:
            milliseconds = round(microseconds / 1000)
            str_value = f"{str_value[:19]}.{milliseconds:03}{str_value[19:]}"

        return cls(str_value)

    def __repr__(self) -> str:
        return f"Timestamp({super().__repr__()})"

    def datetime(self) -> datetime:
        return datetime.fromisoformat(self)


# Keeping mypy and strawberry and pydantic happy:

_Timestamp.__name__ = "Timestamp"

if TYPE_CHECKING:
    Timestamp = Annotated[TimestampParsable, _Timestamp]
else:
    Timestamp = _Timestamp
